### ISSUE TITLE
Memory leaks when running unit tests

### ISSUE THREAD

**Issue by rffrancon** [rmcolq/pandora#65]
Valgrind output: [valgrind_output.txt](https://github.com/rmcolq/pandora/files/2487664/valgrind_output.txt)

Valgrind command used:
`valgrind --log-file="valgrind_output" ../../cmake-build-debug/test/pandora_test`
with working directory: `pandora/test/test_cases`. This command was executed on my laptop. All of the tests passed.

Commit: b914159c6ab01d3818aa731f37618038a5a39487
Branch: `dev`

For example:
```
==15934== Process terminating with default action of signal 6 (SIGABRT)
==15934==    at 0x5DF9E97: raise (raise.c:51)
==15934==    by 0x5DFB800: abort (abort.c:79)
==15934==    by 0x5DEB399: __assert_fail_base (assert.c:92)
==15934==    by 0x5DEB411: __assert_fail (assert.c:101)
==15934==    by 0x5E29E9: Interval::Interval(unsigned int, unsigned int) (interval.cpp:11)
==15934==    by 0x6136FD: Minimizer::Minimizer(unsigned long, unsigned int, unsigned int, bool) (minimizer.cpp:11)
==15934==    by 0x71C536: MinimizerTest_create_Test::TestBody() (minimizer_test.cpp:51)
==15934==    by 0x800F76: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (in /home/rffrancon/github/pandora/cmake-build-debug/test/pandora_test)
==15934==    by 0x7FAEA2: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (in /home/rffrancon/github/pandora/cmake-build-debug/test/pandora_test)
==15934==    by 0x7D99E9: testing::Test::Run() (in /home/rffrancon/github/pandora/cmake-build-debug/test/pandora_test)
==15934==    by 0x7DA346: testing::TestInfo::Run() (in /home/rffrancon/github/pandora/cmake-build-debug/test/pandora_test)
==15934==    by 0x7DA9C9: testing::TestCase::Run() (in /home/rffrancon/github/pandora/cmake-build-debug/test/pandora_test)
==15934== 
==15934== HEAP SUMMARY:
==15934==     in use at exit: 7,154,922 bytes in 24,129 blocks
==15934==   total heap usage: 571,866 allocs, 547,737 frees, 125,362,426,822 bytes allocated
==15934== 
==15934== LEAK SUMMARY:
==15934==    definitely lost: 5,792 bytes in 71 blocks
==15934==    indirectly lost: 2,722,572 bytes in 14,408 blocks
==15934==      possibly lost: 1,996,141 bytes in 931 blocks
==15934==    still reachable: 2,430,417 bytes in 8,719 blocks
==15934==         suppressed: 0 bytes in 0 blocks
==15934== Rerun with --leak-check=full to see details of leaked memory
```

**Comment by rffrancon** [comment]
Running the same unit tests on the yoda cluster produces a seg fault.

**Comment by rmcolq** [comment]
I can compile pandora and tests and run the tests on yoda on my `dev_fix_vcf_ref` which isn't many commits away from `dev`. Which unit tests segfault?

Could there be places where I have used `new` to set up tests, and not `delete` (or an equivalent) at the end of a test?

**Comment by rffrancon** [comment]
Most leaks are "still reachable" leaks. Other leaks are no longer reachable but occur in GATB code. Here are the lost leaks only (non-"still reachable"): [valgrind_lost_leak_only.txt](https://github.com/rmcolq/pandora/files/2488018/valgrind_lost_leak_only.txt)

Here's one example of a memory leak in unit tests:
Consider this use of `new`:
https://github.com/rmcolq/pandora/blob/756d423a69d9eb7ec5224452191e7c87fbbf7760/src/utils.cpp#L131

It's called in a unit test here:
https://github.com/rmcolq/pandora/blob/756d423a69d9eb7ec5224452191e7c87fbbf7760/test/localPRG_test.cpp#L1151

But I don't think it's ever freed.

**Comment by rmcolq** [comment]
So solution is to either delete all the pointers in the `vector<LocalPRG*> prgs` at the end of the test, or to replace the pointer with some sort of shared pointer in that function? What would you suggest?

**Comment by rffrancon** [comment]
A reliable quick fix would be to replace with shared pointers.

**Comment by rffrancon** [comment]
Test which fails on yoda: `ExtractReadsTest.find_interval_in_localpath_minimal`.

**Comment by mbhall88** [comment]
Is this an issue still?

**Comment by leoisl** [comment]
Memory leaks are still an issue, however I found them in real and toy datasets. They were not all solved with smart pointers due to circular dependency issues with smart pointers (https://www.learncpp.com/cpp-tutorial/15-7-circular-dependency-issues-with-stdshared_ptr-and-stdweak_ptr/). I am trying to solve memory leaks in `pandora compare` on https://github.com/leoisl/pandora/commit/e6cac5339ac5c7a68a636058700a295ebb08d2b9, https://github.com/leoisl/pandora/commit/6d8004805909b1ac4e678faa994163904dbe7b49 and https://github.com/leoisl/pandora/commit/5d4ad0639d13fd3fd07b76670d1bb94c6de09b42 (these commits have more details on how these memory leaks look like). This is the current heap/leak summary from `valgrind memcheck` on a toy dataset in `pandora compare`: 
[5d4ad0_heap_leak_summary.txt](https://github.com/rmcolq/pandora/files/3272922/5d4ad0_heap_leak_summary.txt)


This seems fine - the 4 largest memory leaks we have now are due to OpenMP and Boost logging (I don't really know why we have them, but they seem very minor anyway, and I don't think we can do anything in this case...), but we need to pay attention if there are still some memory leaks in `pandora compare` (my runs are not covering all `compare` code I am afraid) and we need also to check for leaks in  `pandora map` and `index`.

**Comment by mbhall88** [comment]
When you say there is a memory leak in Boost logging, how much are we talking? I don't get how logging could be cause memory issues, it should just be dumping things to stderr?  
Are you sure it just isn't related to [this](https://stackoverflow.com/questions/26542618/the-use-of-boost-log-valgrind-check-memory-leak)? i.e it is more a problem with valgrind being overzealous?

**Comment by leoisl** [comment]
I was wondering also how logging could make memory leaks... But you are right, it is just valgrind being overzealous. Most of it is "still reachable" leaks happening in Boost/OpenMP, and some are possibly lost, so I think we are fine:
```
==2401== LEAK SUMMARY:
==2401==    definitely lost: 0 bytes in 0 blocks
==2401==    indirectly lost: 0 bytes in 0 blocks
==2401==      possibly lost: 2,128 bytes in 7 blocks
==2401==    still reachable: 10,136 bytes in 68 blocks
==2401==         suppressed: 0 bytes in 0 blocks
```

**Comment by leoisl** [comment]
For completeness, this shows that we are indeed fine. This is the leak summary using only **1 thread**:

```
==8636== LEAK SUMMARY:
==8636==    definitely lost: 0 bytes in 0 blocks
==8636==    indirectly lost: 0 bytes in 0 blocks
==8636==      possibly lost: 0 bytes in 0 blocks
==8636==    still reachable: 720 bytes in 8 blocks
==8636==         suppressed: 0 bytes in 0 blocks
```

Now we have 0 "possibly lost" and very few "still reachable", which is usually not a problem.

When using more threads with OpenMP, `valgrind` will see pseudo-memory leaks which we should not worry (see https://medium.com/@auraham/pseudo-memory-leaks-when-using-openmp-11a383cc4cf9 ).

So I guess we can say no memory leaks in `pandora compare` anymore.

**Comment by mbhall88** [comment]
Is this also the case for unit tests (I.e the title of this issue)?

**Comment by leoisl** [comment]
No, I still have to adapt the unit tests to the interfaces I changed (not compiling right now) and do the test. It is also true that this is an overstatement:

> So I guess we can say no memory leaks in pandora compare anymore.

I just did not get memory leaks in one test dataset, which I am sure it does not cover the whole `pandora compare` code...
